src/lib/homestore/shard_manager.cpp

```diff
@@ -20,4 +124,74 @@ ShardManager::Result< InfoList > HSHomeObject::_list_shards(pg_id pg) const {
     return folly::makeUnexpected(ShardError::UNKNOWN_PG);
 }
 
+void HSHomeObject::on_shard_message_commit(int64_t lsn, sisl::blob const& header, sisl::blob const& key,
+                                           void* user_ctx) {
+    const ReplicationMessageHeader* msg_header = r_cast<const ReplicationMessageHeader*>(header.bytes);
+    folly::Promise<ShardManager::Result<ShardInfo> > *promise = nullptr;
+    // user_ctx will be nullptr when:
+    // 1. on the follower side
+    // 2. on the leader side but homeobject restarts and replay all commited log entries from the last commit id;
+    if (user_ctx != nullptr) {
+        promise = r_cast<folly::Promise<ShardManager::Result<ShardInfo> >*> (user_ctx);      
+    }
+
+    if (msg_header->header_crc != msg_header->calculate_crc()) {
+        LOGWARN("replication message header is corrupted with crc error, lsn:{}", lsn);
+        if (promise) {
+            promise->setValue(folly::makeUnexpected(ShardError::INVALID_ARG));	  
+        }
+        return;
+    }
+
+    std::string string_value;
+    string_value.append(r_cast<char*>(header.bytes + sizeof(ReplicationMessageHeader)), msg_header->payload_size);
+
+    auto crc = crc32_ieee(init_crc32, r_cast< const uint8_t* >(string_value.c_str()), string_value.size());
+    if (msg_header->payload_crc != crc) {
+        LOGWARN("replication message body is corrupted with crc error, lsn:{}", lsn);
+        if (promise) {
+            promise->setValue(folly::makeUnexpected(ShardError::INVALID_ARG));	  
+        }
+        return;
+    }
+
+    //TODO: preapre ShardInfoExt msg with chunk_id and persisted the ShardInfoExt to homestore MetaBlkService;
+    void* cookie = nullptr;
```

Comment:
where is cookie saved? how to get cookie when restarting

Reply:
cookie is not saved any more, it is usually to save cookie when you need to update homestore meta blks, but in our cases,   this meta blk will not be updated until now. when shard is sealed,  it will write to a new meta blks to homestore according to homestore design doc.

Reply:
I think metablk for this shard will be deleted too if while shard is deleted. for now, the cookie is saved in shard and can be used to handle to delete this shard info from homestore metablk service.